tools/watermill-cli/cmd/root.go

```diff
@@ -0,0 +1,147 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/spf13/pflag"
+
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var cfgFile string
+var logger watermill.LoggerAdapter
+
+var rootCmd = &cobra.Command{
+	Use:   "watermill",
+	Short: "A CLI for Watermill.",
+	Long: `A CLI for Watermill.
+
+Use console-based producer or consumer for various pub/sub providers.`,
+	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+		if viper.GetBool("log") {
+			logger = watermill.NewStdLogger(viper.GetBool("debug"), viper.GetBool("trace"))
+		} else {
+			logger = watermill.NopLogger{}
+		}
+
+		if err := checkRequiredFlags(cmd.Flags()); err != nil {
+			return err
+		}
+
+		writeConfig := viper.GetString("writeConfig")
+		if writeConfig != "" {
+			settings := viper.AllSettings()
+			delete(settings, "writeconfig")
+			b, err := yaml.Marshal(settings)
+			if err != nil {
+				return errors.Wrap(err, "could not marshal config to yaml")
+			}
+
+			f, err := os.Create(writeConfig)
+			if err != nil {
+				return errors.Wrap(err, "could not create file for write")
+			}
+			_, err = fmt.Fprintf(f, "%s", b)
+			if err != nil {
+				return errors.Wrap(err, "could not write to file")
+			}
+		}
+
+		return nil
+	},
+}
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+	rootCmd.PersistentFlags().SortFlags = false
+
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.watermill-cli.yaml)")
+
+	outputFlags := pflag.NewFlagSet("output", pflag.ExitOnError)
+	outputFlags.BoolP("log", "l", false, "If true, the logger output is sent to stderr. No logger output otherwise.")
+	ensure(viper.BindPFlag("log", outputFlags.Lookup("log")))
+
+	outputFlags.Bool("debug", false, "If true, debug output is enabled from the logger")
```

Comment:
can you add a shor version ("-d"?)

Reply:
I guess could. But I don't think that would be a good idea for `--trace`, because a lot of commands have `-t, --topic`.

We have to watch out for clashes with subcommands' flags

Reply:
b3ae9f252ccb6e90efcf9f72512765a9df71cff4 done. No changes for `--trace`